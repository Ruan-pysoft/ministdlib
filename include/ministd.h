#ifndef _RM_STD_H
#define _RM_STD_H

#include <_ministd_prelude.h>

/* This is the main header for providing basic startup and exit functionality,
 * as well as some basic interaction with the program's environment.
 */

/* These functions are used to access
 * the program's command-line arguments (argc/argv)
 * and the program's environment variables (envc/envp)
 */
int argc(void);
char ref ref argv(void);
int envc(void);
char ref ref envp(void);

/* Functions to terminate the program */
void atexit(void (ref func)(void));
void exit(int exitcode);
void thread_exit(int exitcode);
void panic(isz exitcode);
#define PANIC(code) do { panic(code); __builtin_unreachable(); } while (0)

/* The program's entry point, called by the kernel
 * _start in turn call's the user's main.
 * See https://en.wikipedia.org/wiki/Entry_point#Contemporary */
void _start(void);

#endif /* _RM_STD_H */
