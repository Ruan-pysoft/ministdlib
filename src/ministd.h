#ifndef _RM_STD_H
#define _RM_STD_H

#include "ministd_io.h"
#include "ministd_types.h"

int argc(void);
char ref ref argv(void);
int envc(void);
char ref ref envp(void);

void atexit(void (ref func)(void));
void exit(int exitcode);
void _start(void);

#endif /* _RM_STD_H */
