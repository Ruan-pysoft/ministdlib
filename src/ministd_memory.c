#include <ministd_memory.h>

#include <ministd_error.h>
#include <ministd_syscall.h>

/* screw it, I'll do this manually later */
#define _SYS_MMAN_H
#include <bits/mman-linux.h>

own_ptr mmap(ptr addr, usz len, int prot, int flags, int fildes, usz off, err_t ref err_out);
own_ptr
mmap(ptr addr, usz len, int prot, int flags, int fildes, usz off, err_t ref err_out)
{
	own_ptr res;

	_syscall6(__NR_mmap, res, addr, len, prot, flags, fildes, off);

	if ((isz)res < 0) {
		ERR_WITH(-(isz)res, NULL);
	}

	return res;
}
void munmap(ptr addr, usz len, err_t ref err_out);
void
munmap(ptr addr, usz len, err_t ref err_out)
{
	long res;

	_syscall2(__NR_munmap, res, addr, len);

	if (res < 0) ERR_VOID(-res);
}

own_ptr std_alloc(Allocator ref this, usz bytes, err_t ref err_out);
own_ptr std_realloc(Allocator ref this, own_ptr buf, usz bytes, err_t ref err_out);
void std_free(Allocator ref this, own_ptr buf);
static Allocator std_allocator = {
	(Allocator_alloc_t)std_alloc,     /* alloc */
	(Allocator_realloc_t)std_realloc, /* realloc */
	(Allocator_free_t)std_free,       /* free */
};
typedef char ALIGN[16];
typedef union std_header {
	struct {
		usz size;
		int is_free;
	} s;
	ALIGN stub;
} std_header_t;

own_ptr
std_alloc(Allocator ref this, usz bytes, err_t ref err_out)
{
	err_t err = ERR_OK;
	usz total_size;
	own_ptr block;
	std_header_t ref header;

	(void) this;

	if (!bytes) return NULL;

	total_size = sizeof(std_header_t) + bytes;
	block = mmap(NULL, total_size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0, &err);
	TRY_WITH(err, NULL);
	if (block == NULL) {
		return NULL;
	}

	header = block;
	header->s.size = bytes;
	header->s.is_free = 0;

	return (ptr)(header + 1);
}
own_ptr
std_realloc(Allocator ref this, own_ptr buf, usz bytes, err_t ref err_out)
{
	err_t err = ERR_OK;
	std_header_t ref header;
	own_ptr res;

	if (!buf) return std_alloc(this, bytes, err_out);
	if (!bytes) {
		std_free(this, buf);
		return NULL;
	}
	header = (std_header_t*)buf - 1;
	if (header->s.size >= bytes) return buf;

	res = std_alloc(this, bytes, &err);
	if (err != ERR_OK) {
		std_free(this, buf);
		ERR_WITH(err, NULL);
	}
	if (res) {
		memmove(res, buf, bytes < header->s.size ? bytes : header->s.size);
	}
	std_free(this, buf);

	return res;
}
void
std_free(Allocator ref this, own_ptr buf)
{
	std_header_t ref header;

	(void) this;

	if (!buf) return;

	header = (std_header_t ref)buf - 1;

	munmap((ptr)header, sizeof(std_header_t) + header->s.size, NULL);
}

own_ptr
alloc(usz bytes, err_t ref err_out)
{
	return a_alloc(&std_allocator, bytes, err_out);
}
own_ptr
nalloc(usz size, usz n, err_t ref err_out)
{
	return a_nalloc(&std_allocator, size, n, err_out);
}
own_ptr
realloc(own_ptr buf, usz bytes, err_t ref err_out)
{
	return a_realloc(&std_allocator, buf, bytes, err_out);
}
own_ptr
nrealloc(own_ptr buf, usz size, usz n, err_t ref err_out)
{
	return a_nrealloc(&std_allocator, buf, size, n, err_out);
}
void
free(own_ptr buf)
{
	a_free(&std_allocator, buf);
}

void
memzero(ptr buf, usz bytes)
{
	usz i;

	for (i = 0; i < bytes; ++i) {
		((char ref)buf)[i] = 0;
	}
}
void
nmemzero(ptr buf, usz size, usz n, err_t ref err_out)
{
	usz total_size;

	if (!size || !n) return;
	total_size = size * n;
	if (total_size / n != size) {
		ERR_VOID(ERR_OVERFLOW);
	}
	memzero(buf, total_size);
}
void
memmove(void ref dest, const void ref src, usz n)
{
	usz i;
	for (i = 0; i < n; ++i) {
		((char ref)dest)[i] = ((char ref)src)[i];
	}
}
void
nmemmove(void ref dest, const void ref src, usz size, usz n, err_t ref err_out)
{
	usz total_size;

	if (!size || !n) return;
	total_size = size * n;
	if (total_size / n != size) {
		ERR_VOID(ERR_OVERFLOW);
	}
	memmove(dest, src, total_size);
}

own_ptr sbrk(isz n, err_t ref err_out);
own_ptr
sbrk(isz n, err_t ref err_out)
{
	ptr increment_to, curr_limit;
	own_ptr res;

	_syscall1(__NR_brk, curr_limit, 0); /* get current program break */

	if (n == 0) return curr_limit;

	increment_to = (char*)curr_limit + n;
	if (increment_to < curr_limit) { /* overflow */
		ERR_WITH(ERR_NOMEM, NULL);
	}

	_syscall1(__NR_brk, res, increment_to); /* increment program break */

	if (res == (ptr)-1) {
		/* NOTE: this is my best guess as to what the syscall does
		 * based off of the libc manpages,
		 * but I'm unsure;
		 * TODO: verify */
		ERR_WITH(ERR_NOMEM, NULL);
	}

	return (char*)res - n; /* want old value, not new value */
}

own_ptr sbrk_alloc(Allocator ref this, usz bytes, err_t ref err_out);
own_ptr sbrk_realloc(Allocator ref this, own_ptr buf, usz bytes, err_t ref err_out);
void sbrk_free(Allocator ref this, own_ptr buf);
Allocator sbrk_allocator = {
	(Allocator_alloc_t)sbrk_alloc,
	(Allocator_realloc_t)sbrk_realloc,
	(Allocator_free_t)sbrk_free,
};
typedef union sbrk_header {
	struct {
		usz size;
		int is_free;
		union sbrk_header ref next;
	} s;
	ALIGN stub;
} sbrk_header_t;
sbrk_header_t ref sbrk_head, ref sbrk_tail;

/* FIXME: I'm pretty sure there's various bugs lurking in the sbrk_* suite,
 * which I found & patched while translating it to a mmap version,
 * but they're still hanging out in here...
 */
sbrk_header_t ref sbrk_get_free_block(usz bytes);
own_ptr
sbrk_alloc(Allocator ref this, usz bytes, err_t ref err_out)
{
	err_t err = ERR_OK;
	usz total_size;
	own_ptr block;
	sbrk_header_t ref header;

	(void) this;

	if (!bytes) return 0;

	header = sbrk_get_free_block(bytes);
	if (header) {
		header->s.is_free = 0;
		return (ptr)(header + 1);
	}

	total_size = sizeof(sbrk_header_t) + bytes;
	block = sbrk(total_size, &err);
	TRY_WITH(err, NULL);

	header = block;
	header->s.size = bytes;
	header->s.is_free = 0;
	header->s.next = 0;
	if (!sbrk_head) sbrk_head = header;
	if (sbrk_tail) sbrk_tail->s.next = header;
	sbrk_tail = header;

	return (ptr)(header + 1);
}
sbrk_header_t ref
sbrk_get_free_block(usz bytes)
{
	sbrk_header_t ref res;

	for (res = sbrk_head; res != 0; res = res->s.next) {
		if (res->s.is_free && res->s.size >= bytes) return res;
	}

	return res;
}
own_ptr
sbrk_realloc(Allocator ref this, own_ptr buf, usz bytes, err_t ref err_out)
{
	err_t err = ERR_OK;
	sbrk_header_t ref header;
	own_ptr ret;

	if (!buf || !bytes) return sbrk_alloc(this, bytes, err_out);
	header = (sbrk_header_t*)buf - 1;
	if (header->s.size >= bytes) return buf;
	ret = sbrk_alloc(this, bytes, &err);
	TRY_WITH(err, ret);
	if (ret) {
		usz i;

		for (i = 0; i < header->s.size; ++i) {
			((char*)ret)[i] = ((char*)buf)[i];
		}
	}
	sbrk_free(this, buf);

	return ret;
}
void
sbrk_free(Allocator ref this, own_ptr buf)
{
	sbrk_header_t ref header, ref tmp;
	ptr programbreak;

	(void) this;

	if (!buf) return;
	header = ((sbrk_header_t*)buf) - 1;

	programbreak = sbrk(0, NULL);
	if (programbreak == NULL) {
		perror(ERR_NOMEM, "Something went wrong when calling free");
		return;
	}
	if ((char*)buf + header->s.size == programbreak) {
		if (sbrk_head == sbrk_tail) {
			sbrk_head = sbrk_tail = 0;
		} else {
			int counter = 0;

			tmp = sbrk_head;
			while (tmp && counter < (1<<16)) {
				if (tmp->s.next == sbrk_tail) {
					tmp->s.next = 0;
					sbrk_tail = tmp;
				}
				tmp = tmp->s.next;

				/* quick-and-dirty infinite loop protection */
				/* (but you'll probably SEGFAULT before an infinite loop) */
				++counter;
			}
		}
		if (sbrk(0 - sizeof(sbrk_header_t) - header->s.size, NULL) == NULL) {
			perror(ERR_NOMEM, "Something went wrong when calling free");
			return;
		}
		return;
	}
	header->s.is_free = 1;
}

own_ptr
a_alloc(Allocator ref this, usz bytes, err_t ref err_out)
{
	return this->alloc(this, bytes, err_out);
}
own_ptr
a_nalloc(Allocator ref this, usz size, usz n, err_t ref err_out)
{
	usz total_size;

	if (!size || !n) return NULL;
	total_size = size * n;
	if (total_size / n != size) {
		ERR_WITH(ERR_OVERFLOW, NULL);
	}
	return a_alloc(this, total_size, err_out);
}
own_ptr
a_realloc(Allocator ref this, own_ptr buf, usz bytes, err_t ref err_out)
{
	return this->realloc(this, buf, bytes, err_out);
}
own_ptr
a_nrealloc(Allocator ref this, own_ptr buf, usz size, usz n, err_t ref err_out)
{
	usz total_size;

	if (!size || !n) {
		free(buf);
		return NULL;
	}
	total_size = size * n;
	if (total_size / n != size) {
		free(buf);
		ERR_WITH(ERR_OVERFLOW, NULL);
	}
	return a_realloc(this, buf, total_size, err_out);
}
void
a_free(Allocator ref this, own_ptr buf)
{
	this->free(this, buf);
}
