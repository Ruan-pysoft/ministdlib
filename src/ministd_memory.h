#ifndef _RM_STD_MEMORY_H
#define _RM_STD_MEMORY_H

#include "ministd_types.h"

ptr alloc(usz);
ptr nalloc(usz, usz);
ptr realloc(ptr, usz);
ptr nrealloc(ptr, usz, usz);
void free(ptr);

void memzero(ptr, usz);
void nmemzero(ptr, usz, usz);

typedef struct Allocator {
	ptr (*alloc)(struct Allocator ref, usz);
	ptr (*realloc)(struct Allocator ref, ptr, usz);
	void (*free)(struct Allocator ref, ptr);
} Allocator;

ptr a_alloc(Allocator ref, usz);
ptr a_nalloc(Allocator ref, usz, usz);
ptr a_realloc(Allocator ref, ptr, usz);
ptr a_nrealloc(Allocator ref, ptr, usz, usz);
void a_free(Allocator ref, ptr);

extern Allocator ref stack_allocator;

Allocator ref a_newarena(ptr, usz);

#endif /* _RM_STD_MEMORY_H */
