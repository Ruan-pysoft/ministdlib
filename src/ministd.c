#include "ministd.h"

#include "ministd_syscall.h"

#include <asm/unistd.h>

void
exit(int exitcode)
{
	isz lcode, dummy;
	lcode = exitcode;

	_syscall1(__NR_exit, dummy, lcode);
	__builtin_unreachable();
}
int main(int argc, char ref ref argv);
extern void ministd_io_init(void);
void
setup(void)
{
	ministd_io_init();
}
void
_start(void)
{
	int argc;
	char ref ref argv;

	__asm__ volatile(
		"mov %%rsp, %%rax;" /* sp -> ax */
		"mov 24(%%rax), %[argc];" /* retrieve argc, top item of stack */
		"mov %%rax, %[argv];" /* retrieve argv, second-to-top of stack */
		"add $32, %[argv];"
		: [argc] "=r" (argc), [argv] "=r" (argv) /* outputs */
		: /* inputs */
		: "rax" /* clobber */
	);

	setup();

	exit(main(argc, argv));
	__builtin_unreachable();
}

#include "ministd_io.h"

void
__stack_chk_fail(void)
{
	const char msg[] = "Stack smashing detected!";

	fd_write(2 /* stderr */, (ptr)msg, sizeof(msg)/sizeof(*msg), NULL);
	exit(127);
}
