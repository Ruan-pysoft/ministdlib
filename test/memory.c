#include "ministd.h"
#include "ministd_memory.h"

char *allocs[1024];
const char *test_string = "Hai #.... :3";

int
strlen(const char ref str)
{
	int res;

	for (res = 0; str[res] != 0; ++res);

	return res;
}

int
main(void)
{
	err_t err = ERR_OK;
	int i;

	puts("Starting memory test...", NULL);

	for (i = 0; i < 1024; ++i) {
		int j;

		allocs[i] = alloc(16, &err);
		if (err) {
			perror(err, "Something went wrong while allocating memory");
			exit(127);
		}
		for (j = 0; j < strlen(test_string); ++j) {
			allocs[i][j] = test_string[j];
		}
		allocs[i][5] = '0' + (i/1000)%10;
		allocs[i][6] = '0' + (i/100)%10;
		allocs[i][7] = '0' + (i/10)%10;
		allocs[i][8] = '0' + i%10;
	}

	for (i = 0; i < 1024; i += 128) {
		puts(allocs[i], NULL);
	}
	puts(allocs[13], NULL);
	puts(allocs[1023], NULL);

	for (i = 0; i < 1024; ++i) {
		free(allocs[i]);
	}

	return 0;
}
