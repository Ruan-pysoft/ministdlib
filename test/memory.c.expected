Starting memory test...
Hai #0000 :3
Hai #0128 :3
Hai #0256 :3
Hai #0384 :3
Hai #0512 :3
Hai #0640 :3
Hai #0768 :3
Hai #0896 :3
Hai #0013 :3
Hai #1023 :3
